packages/ui/src/components/va-stepper/VaStepper.vue

```diff
@@ -1,100 +1,51 @@
 <template>
-  <div
-    class="va-stepper"
-    :class="{ 'va-stepper--vertical': $props.vertical }"
-    v-bind="ariaAttributesComputed"
-  >
-    <ol
-      class="va-stepper__navigation"
-      ref="stepperNavigation"
-      :class="{ 'va-stepper__navigation--vertical': $props.vertical }"
+  <VaForm ref="form">
```

Comment:
It feels like Form must be used outside of stepper. Better register each step as form field and add validation method to it.

Reply:
When you say must, what do you mean? Functionally? Would doing it this way cause performance issues?

Can you explain more what you're suggesting? Do you mean that each step should be a va-form? How would you recommend doing that?

Reply:
The interface must be like

```html
<VaForm ref="form">
  <VaStepper>
      ...
  </VaStepper>
</VaForm>
```
It is more tree-shaking efficient, because not all steps requires a form inside. Also, with VaForm used as parent for VaStepper allows user to access form with `useForm`.

Usage must be like:
```vue
<template>
  <VaForm ref="stepperForm">
    <VaStepper v-model="step" :steps="steps" :next-disabled="!isValid" @before-next="(next) => next(validate)">
      <template #step-content-0>
        <va-input
          v-model="model.a"
          :rules="[(v) => v.length > 0 || 'Required']"
          label="A"
        >
        </va-input>
      </template>
      <template #step-content-1>
        <va-input
          v-model="model.b"
          :rules="[(v) => v === '13' || 'Value must be 13']"
          label="B"
        ></va-input>
      </template>
      <template #step-content-2>
        <va-input v-model="model.c" label="C"></va-input>
      </template>
    </VaStepper>
  </VaForm>
</template>

<script setup lang="ts">
import { ref } from "vue";
import { useForm } from "vuestic-ui";

const { isValid, validate } = useForm("stepperForm");

const step = ref();
const model = ref({ a: "", b: "", c: "" });

const steps = [{ label: "One" }, { label: "Two" }, { label: "Three" }];
</script>
```

It does the same thing you have with `$props.linear` and also disables nextButton visually.
The only thing we need to add is event `before-next` where we can validate form (`@before-next="(next) => next(validate)` like vue-router guard). But it is actually not required, because you can pass custom buttons.

Here is a working example: https://codesandbox.io/p/sandbox/vibrant-orla-sqrjc4?file=/src/App.vue:3,46-3,64 (Idk why codesandbox draws controls two times, btw)

Do you feel like you still need this behavior to be done with prop or we can move validation process out of component?

Reply:
One thing to note, in your example navigation is disabled. It would be nice to still retain that functionality

Reply:
@m0ksem can you check out my last wip change? I'm trying to pass form as ref. It does show as a ref, but validate is always undefined. I've tried passing the actual ref and the string name for the form.

Reply:
I'm not sure how to implement validation with navigation enabled...

If you have step1 valid, you must be able to go to step2.
In case you're back to step1 and made it invalid, you mustn't be able to go to step2 again.

Also, VaForm validates only mounted components, so we're not able to check if step1 valid after moved to step2.


---
I think this can be fixed if nextDisabled will disable step2 if step1 is invalid, even if you typed something in form. Looks like correct behavior for me.

Reply:
I most ideal (imo) would be to have a prop for the form on the stepper. If the form is passed, then it knows validation should occur. This would negate the need for linear property. But Vue automatically unwrapping the ref, and then useForm expecting a string or Ref<typeof VaForm> is causing an issue. I'm searching S.O and open issues, but I'm not seeing a resolution. I've tried nesting the formRef inside an object (so it isnt unwrapped) but that did not work. Other suggestions were to use makeRaw but that also does not work. 

I think this type of approach would provide a simpler and thus more enjoyable development experience if possible.

Reply:
I think this would be accomplished by adding a prop of completed on each step. When you successfully validate and then update the current step index, then you also set that step as completed. Then you cannot proceed to a step unless the previous step is completed and the current step is valid. Thoughts? I think this is basically how angular material works. They have a linear mode but you can also use the upper navigation.

Reply:
> I'm not sure how to implement validation with navigation enabled...
> 
> If you have step1 valid, you must be able to go to step2. In case you're back to step1 and made it invalid, you mustn't be able to go to step2 again.
> 
> Also, VaForm validates only mounted components, so we're not able to check if step1 valid after moved to step2.
> 
> I think this can be fixed if nextDisabled will disable step2 if step1 is invalid, even if you typed something in form. Looks like correct behavior for me.

I think this would be accomplished by adding a prop of completed on each step. When you successfully validate and then update the current step index, then you also set that step as completed. Then you cannot proceed to a step unless the previous step is completed and the current step is valid. Thoughts? I think this is basically how angular material works. They have a linear mode but you can also use the upper navigation.

Reply:
It looks like just didn't register VaForm in demo component. 
Also, notice that you mustn't use `useForm` in methods, only during setup fn execution. You can directly access validate method in VaForm ref. 
You also don't need to use `FormRef`, even thought vue unwrapps ref in props `props.form` is still reactive, it is because props are`reactive`. 
```ts
const form = toRef(props, 'form')
```
`form` will not lose reactivity this way.

In template, all refs are unwrapped, so in case `<VaStepper :form="myCustomForm" />` will be compiled like
```ts
h(VaStepper, {
   form: myCustomForm.value,
})
```
(actually it will use ctx.myCustomForm which is reactive too, so code above is just analogy)

Reply:
Okay cool. That does work exactly as intended! Sorry, I'm still getting a total hang of Vue's reactivity.

Reply:
@m0ksem after playing around a bit with it, I did opt to not have the stepper trigger form validation automatically. You of course can still wrap the stepper with the form and then complete the step or set the step as errored if validation fails.